password-cli/app/Main.hs

```diff
@@ -0,0 +1,209 @@
+            <> command "bcrypt" (info (pure BcryptCheckAlgo) (progDesc "Bcrypt"))
+            <> command "pbkdf2" (info (pure PBKDF2CheckAlgo) (progDesc "PBKDF2"))
+            <> command "scrypt" (info (pure ScryptCheckAlgo) (progDesc "Scrypt"))
+        )
+        <**> helper
+
+data AlgorithmeDef a p = AlgorithmeDef
+  { algoParser :: Parser p
+  , algoHash :: p -> Password -> IO (PasswordHash a)
+  , algoCheck :: Password -> PasswordHash a -> PasswordCheck
+  }
+
+argon2Def :: AlgorithmeDef Argon2.Argon2 Argon2.Argon2Params
+argon2Def = AlgorithmeDef
+  { algoParser = pure Argon2.defaultParams
+  , algoHash = Argon2.hashPasswordWithParams
+  , algoCheck = Argon2.checkPassword
+  }
+
+bcryptDef :: AlgorithmeDef Bcrypt.Bcrypt Int
+bcryptDef = AlgorithmeDef
+  { algoParser = pure Bcrypt.defaultParams
+  , algoHash = Bcrypt.hashPasswordWithParams
+  , algoCheck = Bcrypt.checkPassword
+  }
+
+pbkdf2Def :: AlgorithmeDef PBKDF2.PBKDF2 PBKDF2.PBKDF2Params
+pbkdf2Def = AlgorithmeDef
+  { algoParser = pure PBKDF2.defaultParams
+  , algoHash = PBKDF2.hashPasswordWithParams
+  , algoCheck = PBKDF2.checkPassword
+  }
+
+scryptDef :: AlgorithmeDef Scrypt.Scrypt Scrypt.ScryptParams
+scryptDef = AlgorithmeDef
+  { algoParser = pure Scrypt.defaultParams
+  , algoHash = Scrypt.hashPasswordWithParams
+  , algoCheck = Scrypt.checkPassword
+  }
+
+runCmd :: Cmd -> IO ()
+runCmd =
+  \case
+    HashCmd opts -> runHashCmd opts
+    CheckCmd opts -> runCheckCmd opts
+    HelpCmd mCmd -> runHelpCmd mCmd
+
+runHashCmd :: HashOpts -> IO ()
+runHashCmd HashOpts {..} = do
+  pw <- getPassword quiet password
+  hash <-
+        case hashAlgorithm of
+            Argon2HashAlgo p -> unPasswordHash <$> algoHash argon2Def p pw
+            BcryptHashAlgo p -> unPasswordHash <$> algoHash bcryptDef p pw
+            PBKDF2HashAlgo p -> unPasswordHash <$> algoHash pbkdf2Def p pw
+            ScryptHashAlgo p -> unPasswordHash <$> algoHash scryptDef p pw
+  T.putStr hash
+
+runCheckCmd :: CheckOpts -> IO ()
+runCheckCmd CheckOpts {..} = do
+  pw <- getPassword quiet password
+  checked <-
+        case checkAlgorithm of
+            Argon2CheckAlgo -> algoCheck argon2Def pw <$> getHash hash
+            BcryptCheckAlgo -> algoCheck bcryptDef pw <$> getHash hash
+            PBKDF2CheckAlgo -> algoCheck pbkdf2Def pw <$> getHash hash
+            ScryptCheckAlgo -> algoCheck scryptDef pw <$> getHash hash
+  case checked of
+    PasswordCheckSuccess ->
+      T.putStrLn "Hash and password match"
+    PasswordCheckFail -> do
+      T.hPutStrLn stderr "Hash and password do not match"
+      exitFailure
+
+runHelpCmd :: Maybe String -> IO ()
+runHelpCmd mCmd =
+  let args = maybe id (:) mCmd ["-h"]
+   in void $ handleParseResult $ execParserPure defaultPrefs cliOpts args
+
+getPassword :: Bool -> Maybe FilePath -> IO Password
+getPassword quiet =
+  \case
+    Just path -> do
+      when quiet $ do
+        T.hPutStrLn stderr "Can't use --password-file and -q at the same time"
```

Comment:
I'd like to always allow `--quiet`. I don't feel it makes sense from a user perspective to stop the program, just because someone doesn't want any logging to the terminal.